TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+
+
+#pragma mark - Helper methods
+
+- (void)checkObject:(id)object isOfClass:(Class)className allowNil:(BOOL)allowNil
+{
+
+    if (allowNil && !object) {
+        return;
+    }
+
+    if ([object isKindOfClass:className]) {
+        return;
+    }
+
+    NSString *reason = [NSString stringWithFormat:@"Invalid type encountered. Expected object of class %@.", NSStringFromClass(className)];
+    [self failWithErrorCode:TWTJSONSchemaParserErrorCodeInvalidClass reason:reason object:object];
+}
+
+
+- (void)checkObjectIsArrayWithAtLeastOneItem:(id)object allowNil:(BOOL)allowNil
+{
+    [self checkObject:object isOfClass:[NSArray class] allowNil:allowNil];
+    if (!object || [object count] >= 1) {
+        return;
+    }
+
+    NSString *reason = @"Expected array with at least one item.";
+    [self failWithErrorCode:TWTJSONSchemaParserErrorCodeRequiresAtLeastOneItem reason:reason object:object];
+}
+
+
+- (void)checkObject:(id)object isOneOf:(NSSet *)validValues
+{
+    if ([validValues containsObject:object]) {
+        return;
+    }
+
+    NSString *reason = [NSString stringWithFormat:@"Invalid valid encountered. Expected one of %@", validValues];
+    [self failWithErrorCode:TWTJSONSchemaParserErrorCodeInvalidValue reason:reason object:object];
+}
+
+
+- (void)failWithErrorCode:(NSUInteger)code reason:(NSString *)reason object:(id)object
+{
+    NSString *pathString = [@"/" stringByAppendingString:[self.pathStack componentsJoinedByString:@"/"]];
+    object = object ?: [NSNull null];
+    reason = reason ?: @"";
+
+    NSError *error = [NSError errorWithDomain:kTWTJSONSchemaParserErrorInfoDomainKey code:code userInfo:@{ kTWTJSONSchemaParserErrorInfoPathKey : pathString,
+                                                                                                           kTWTJSONSchemaParserErrorInfoInvalidObjectKey : object,
+                                                                                                           NSLocalizedFailureReasonErrorKey : reason }];
+
+    [[NSException exceptionWithName:@"Invalid JSON Schema"
+                             reason:nil
+                           userInfo:@{ kTWTJSONExceptionErrorKey : error }] raise];
+}
+
+
+- (NSSet *)validJSONTypeKeywords
+{
+    static NSSet *types = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        types = [NSSet setWithObjects:kTWTJSONSchemaTypeKeywordAny,
+                 kTWTJSONSchemaTypeKeywordArray,
+                 kTWTJSONSchemaTypeKeywordBoolean,
+                 kTWTJSONSchemaTypeKeywordInteger,
+                 kTWTJSONSchemaTypeKeywordNull,
+                 kTWTJSONSchemaTypeKeywordNumber,
+                 kTWTJSONSchemaTypeKeywordObject,
+                 kTWTJSONSchemaTypeKeywordString,
+                 nil];
+    });
+
+    return types;
+}
+
+
+- (void)pushPathComponent:(id)object
+{
+    [self.pathStack addObject:object];
+}
+
+
+- (void)popPathComponent
+{
+    [self.pathStack removeLastObject];
```

Comment:
Go ahead and return the last object here. You might not use it now, but that’s typically how pop works, so doing so is unexpected.